### PR TITLE
WIP: Add more boundary checks when parsing the ELF relocs ##bin

### PR DESCRIPTION
**Pull request by trufae**
**Your checklist for this pull request**
- [x] I've read the [guidelines for contributing](https://github.com/radareorg/radare2/blob/master/DEVELOPERS.md) to this repository
- [x] I made sure to follow the project's [coding style](https://github.com/radareorg/radare2/blob/master/DEVELOPERS.md#code-style)
- [ ] I've added tests that prove my fix is effective or that my feature works (if possible)
- [ ] I've updated the documentation and the [radare2 book](https://github.com/radareorg/radare2book) with the relevant information (if needed)

**Detailed description**

This PR aims to improve the boundary checks needed to parse relocs in corrupted files. Don't review this PR yet. its just reminiscense of changes wirtten beofre someone else pushed the fixes into master, so i need more time to review

**Test plan**

wait

**Closing issues**

none

### CODE REVIEW

File: libr/bin/format/elf/elf.c [hunk 1]
```diff
@@ -2660,35 +2660,41 @@ static size_t get_num_relocs_sections(ELFOBJ *bin) {
 }
 
 static size_t get_num_relocs_approx(ELFOBJ *bin) {
-	return get_num_relocs_dynamic (bin) + get_num_relocs_sections (bin);
+	size_t rd = get_num_relocs_dynamic (bin);
```
Comment (ret2libc) [inline libr/bin/format/elf/elf.c:2663]: is it really necessary to split?

File: libr/bin/format/elf/elf.c [hunk 2]
```diff
@@ -2660,35 +2660,41 @@ static size_t get_num_relocs_sections(ELFOBJ *bin) {
...
+	size_t p = *pos;
 
-	for (offset = 0; offset < bin->dyn_info.dt_pltrelsz && pos < num_relocs; offset += size, pos++) {
+	for (offset = 0; offset < bin->dyn_info.dt_pltrelsz && p < num_relocs; offset += size) {
 		if (!read_reloc (bin, relocs + pos, bin->dyn_info.dt_pltrel, bin->dyn_info.dt_jmprel + offset)) {
 			break;
 		}
-		fix_rva_and_offset_exec_file (bin, relocs + pos);
+		fix_rva_and_offset_exec_file (bin, relocs + p);
+		i++;
```
Comment (ret2libc) [inline libr/bin/format/elf/elf.c:2678]: what is `i`?

File: libr/bin/format/elf/elf.c [hunk 3]
```diff
@@ -2660,35 +2660,41 @@ static size_t get_num_relocs_sections(ELFOBJ *bin) {
...
+		p++;
```
Comment (ret2libc) [inline libr/bin/format/elf/elf.c:2679]: why not directly in the `for-loop` as it was before?

File: libr/bin/format/elf/elf.c [hunk 4]
```diff
@@ -2660,35 +2660,41 @@ static size_t get_num_relocs_sections(ELFOBJ *bin) {
...
 	}
 
-	for (offset = 0; offset < bin->dyn_info.dt_relasz && pos < num_relocs; offset += bin->dyn_info.dt_relaent, pos++) {
+	for (offset = 0; offset < bin->dyn_info.dt_relasz && p < num_relocs; offset += bin->dyn_info.dt_relaent) {
 		if (!read_reloc (bin, relocs + pos, DT_RELA, bin->dyn_info.dt_rela + offset)) {
 			break;
 		}
-		fix_rva_and_offset_exec_file (bin, relocs + pos);
+		fix_rva_and_offset_exec_file (bin, relocs + p);
+		p++;
```
Comment (ret2libc) [inline libr/bin/format/elf/elf.c:2687]: same

File: libr/bin/format/elf/elf.c [hunk 5]
```diff
@@ -2712,42 +2718,63 @@ static size_t get_next_not_analysed_offset(ELFOBJ *bin, size_t section_vaddr, si
...
-		if (!is_reloc_section (rel_mode) || bin->g_sections[i].size > bin->size || bin->g_sections[i].offset > bin->size) {
+	size_t si = 0; // section index
+	for (si = 0; !bin->g_sections[si].last; si++) {
+		RBinElfSection *s = &bin->g_sections[si];
+		Elf_(Xword) rel_mode = get_section_mode (bin, si);
+		if (!is_reloc_section (rel_mode) || s->size > bin->size || s->offset > bin->size) {
 			continue;
 		}
-
+		if (strstr (s->name, "rela")) {
```
Comment (ret2libc) [inline libr/bin/format/elf/elf.c:2735]: this is wrong. If the function startswith with `.rela` `get_section_mode` would return `DT_RELA` and `is_reloc_section` would return true already. Was there any case where this was necessary?

File: libr/bin/format/elf/elf.c [hunk 6]
```diff
@@ -2712,42 +2718,63 @@ static size_t get_next_not_analysed_offset(ELFOBJ *bin, size_t section_vaddr, si
...
+				break;
+			}
+			fix_rva_and_offset (bin, relocs + p, i);
+			p++;
 		}
+#else
+		ut64 j = get_next_not_analysed_offset (bin, s->offset, 0, bin->baddr);
+		ut64 sa = s->rva;
+		ut64 eos = s->size;
+		while (j != UT64_MAX && p < num_relocs && j < sa) {
```
Comment (ret2libc) [inline libr/bin/format/elf/elf.c:2759]: why checking of UT64_MAX?

Comment (ret2libc) [reply libr/bin/format/elf/elf.c:2759]: why `j < sa`?

File: libr/bin/format/elf/elf.c [hunk 7]
```diff
@@ -2872,16 +2898,24 @@ static RBinElfSection *get_sections_from_phdr(ELFOBJ *bin) {
...
 		return NULL;
 	}
 
-	i = 0;
-	create_section_from_phdr (bin, ret, &i, ".rel.dyn", reldyn, reldynsz);
-	create_section_from_phdr (bin, ret, &i, ".rela.plt", relava, pltgotsz);
-	create_section_from_phdr (bin, ret, &i, ".rel.plt", relva, relasz);
-	create_section_from_phdr (bin, ret, &i, ".got.plt", pltgotva, pltgotsz);
+	size_t i = 0;
+	if (i < num_sections) {
```
Comment (ret2libc) [inline libr/bin/format/elf/elf.c:2907]: this series of `if` doesn't look very good... I'm tempted to pass `num_sections` to `create_section_from_phdr` and let that function check if `i < num_sections`.

File: libr/bin/format/elf/elf.c (hunk 2 above)
Comment (trufae) [reply libr/bin/format/elf/elf.c:2678]: the 9th letter of the abecedary

File: libr/bin/format/elf/elf.c (hunk 1 above)
Comment (trufae) [reply libr/bin/format/elf/elf.c:2663]: imho it's more readable but not necessary. just a matter of taste

**Comment by trufae** [comment]
ill drop the WIP when all tests pass

File: libr/bin/format/elf/elf.c (hunk 1 above)
Comment (ret2libc) [reply libr/bin/format/elf/elf.c:2657]: seems a bit too split... we have much longer lines than this and this imo looks quite readable already
`get_num_relocs_dynamic (bin) + get_num_relocs_sections (bin)`. If this is not ok, I guess that any sort of operation that does not involve super simple type is not.

File: libr/bin/format/elf/elf.c [hunk 8]
```diff
@@ -2706,42 +2708,43 @@ static size_t get_next_not_analysed_offset(ELFOBJ *bin, size_t section_vaddr, si
...
 	if (!bin->g_sections) {
-		return pos;
+		return;
 	}
 
-	for (i = 0; !bin->g_sections[i].last; i++) {
-		rel_mode = get_section_mode (bin, i);
-
-		if (!is_reloc_section (rel_mode) || bin->g_sections[i].size > bin->size || bin->g_sections[i].offset > bin->size) {
+	size_t si = 0; // section index
```
Comment (ret2libc) [inline libr/bin/format/elf/elf.c:2718]: i don't see why this change. we use `i` everywhere and here there doesn't even seem to be confusion with other loops.

File: libr/bin/format/elf/elf.c (hunk 7 above)
Comment (trufae) [reply libr/bin/format/elf/elf.c:2877]: yeah that could be another option... or just use rvector api which already checks for that and remove the original calloc/.last approach

Comment (trufae) [reply libr/bin/format/elf/elf.c:2877]: actually it should never overflow if the num_sections count is done properly. but there's nohting that checks for integer overflows when couting sections above..

File: libr/bin/format/elf/elf.c (hunk 8 above)
Comment (trufae) [reply libr/bin/format/elf/elf.c:2718]: what do you mean?

File: libr/bin/format/elf/elf.c (hunk 1 above)
Comment (trufae) [reply libr/bin/format/elf/elf.c:2657]: i can rollback if you like, i prefer to read in vertical

File: libr/bin/format/elf/elf.c (hunk 5 above)
Comment (trufae) [reply libr/bin/format/elf/elf.c:2725]: I think it was parsing it twice.. but i have to recheck
